androidenhancedvideoplayer/src/main/java/com/profusion/androidenhancedvideoplayer/utils/TrackUtils.kt

```diff
@@ -0,0 +1,99 @@
+import androidx.media3.common.TrackSelectionOverride
+import androidx.media3.common.TrackSelectionParameters
+import androidx.media3.common.util.UnstableApi
+import androidx.media3.exoplayer.trackselection.DefaultTrackSelector
+import androidx.media3.exoplayer.trackselection.MappingTrackSelector
+
+const val AUTO_QUALITY_LABEL = "Auto"
+
+@OptIn(UnstableApi::class)
+fun DefaultTrackSelector.generateTrackQualityOptions():
+    ArrayList<Pair<String, TrackSelectionParameters.Builder>> {
+    val qualityList = ArrayList<Pair<Format, TrackSelectionParameters.Builder>>()
+    val qualityOptions = ArrayList<Pair<String, TrackSelectionParameters.Builder>>()
+    val autoQualityOption = AUTO_QUALITY_LABEL to
+        buildUponParameters()
+            .clearOverridesOfType(C.TRACK_TYPE_VIDEO)
+            .setAllowMultipleAdaptiveSelections(true)
+
+    val renderTrack = this.currentMappedTrackInfo
+    val renderCount = renderTrack?.rendererCount ?: 0
+    for (rendererIndex in 0 until renderCount) {
+        if (isSupportedFormat(renderTrack, rendererIndex)) {
+            val trackGroupType = renderTrack?.getRendererType(rendererIndex)
+            val trackGroups = renderTrack?.getTrackGroups(rendererIndex)
+            val trackGroupsCount = trackGroups?.length!!
+            if (trackGroupType == C.TRACK_TYPE_VIDEO) {
+                for (groupIndex in 0 until trackGroupsCount) {
+                    val videoQualityTrackCount = trackGroups[groupIndex].length
+                    for (trackIndex in 0 until videoQualityTrackCount) {
+                        val isTrackSupported = renderTrack.getTrackSupport(
+                            rendererIndex,
+                            groupIndex,
+                            trackIndex
+                        ) == C.FORMAT_HANDLED
+
+                        if (isTrackSupported) {
+                            val track = trackGroups[groupIndex]
+                            val trackFormat = track.getFormat(trackIndex)
+
+                            if (trackFormat.isVideoTrackPlayable()) {
+                                val trackBuilder = buildUponParameters()
+                                    .clearOverridesOfType(C.TRACK_TYPE_VIDEO)
+                                    .addOverride(
+                                        TrackSelectionOverride(track, listOf(trackIndex))
+                                    )
+                                qualityList += trackFormat to trackBuilder
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    qualityOptions += autoQualityOption
+    qualityOptions += qualityList.sortedWith(
+        compareBy<Pair<Format, TrackSelectionParameters.Builder>>(
+            { it.first.height },
+            { it.first.bitrate }
+        ).reversed()
+    ).map {
+        it.first.buildQualityLabel() to it.second
+    }
+
+    return qualityOptions
+}
```

Comment:
This function is quite lengthy and contains nested loops. Consider breaking it down into smaller, more manageable functions with clear responsibilities. This will make the code more modular and easier to understand.